rfcs/2022--01-06-8596-tcp-source-backpressure.md

```diff
@@ -0,0 +1,160 @@
+
+This removes both drawbacks from Option 1 by having Vector itself dynamically adjust the limit. The question now becomes,
+how do you select that value
+
+The most similar thing already in Vector is ARC (Adaptive Request Concurrency). However, this same algorithm is not
+appropriate for TCP connections. ARC uses the "additive increase / multiplicative decrease" algorithm to frequently adjust
+how many requests are in-flight at the same time. Each request generally has a similar number of events and occur
+frequently enough that the number in-flight can be adjusted quickly.
+
+With TCP connections, it is impossible to know ahead of time how many events a TCP connection will send before accepting it.
+You also don't know how long a connection may live. It could be a static connection that stays open long-term, or it could send just a single event
+then close. It is also generally not possible to forcibly close a connection without dropping events, since most of the
+TCP sources do not have a way to send an "ack" in the protocol.
+
+I believe this option is not feasible. Trying to dynamically limit requests in flight by only choosing when
+to accept a new connection does not give us enough control to both limit the number of in-flight requests
+and also maintain acceptable performance in all cases.
+
+## Option 3 (dynamic request limit)
+
+Instead of trying to control the number of in-flight requests at the connection level, the requests can just be controlled directly.
+The number of TCP connections could stay unlimited, but there is a check before each request is processed that
+can limit how many are actually in-flight. The goal here is to pick a limit to the number of in-flight requests and
+only allow new ones to be processed if you are below that. This can't be limited to an _exact_ number of events, because you
+don't know how many events are in a specific request until it has been decoded. But you can use a limit in the form of
+`x events + y requests`. This would attempt to limit the in-flight events to `x`, plus `y` full requests of events.
+
+The main drawback to this approach is that you have to be able to accept a request before you know how many events
+are in that request, and the total in-flight count can't be updated until after the decoding has finished. A concurrency
+limit must be placed on request decoding (equal to `y` above) in order to limit the overall requests in-flight.
+
+This leads us to the question of what should the value of `x` and `y` be? `x` (the number of events in-flight) can probably
+be set to 0 and ignored, since we already have event buffering in other parts of the system (the pipeline and sink buffers).
+Choosing `y` is balancing performance with memory usage.
+
+I propose that `y` is a value dynamically determined by the request size of previous requests.
+The "request size" will simply be defined as the number of events contained in a request. The running-average of
```

Comment:
We could also consider tweaking how we calculate the running/moving average, similar to choosing the target based on perf tests we could pick an algo based on perf tests too.

Reply:
Might need to build a soak just for this, since the soaks generally don't get into overload territory. Also, exponential weighted moving averages are pretty easy to implement, give good smooth behavior in the case of spiky peers.

Reply:
EWMA is pretty awesome 😆

Reply:
Yes, I was thinking of EWMA when I wrote this. Will probably go with that unless it doesn't work well for some reason.